docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -313,34 +532,124 @@ type LocalObjectTemplate struct {
       Replicas *int `json:"replicas,omitempty"`
     }
     ```
+1.  The `VariableTopology` object represents an instance of a variable.
+    ```golang
+    // VariableTopology can be used to customize the ClusterClass through
+    // patches. It must comply to the corresponding
+    // VariableDefinitionClass defined in the ClusterClass.
+    type VariableTopology struct {
+      // Name of the variable.
+      Name string `json:"name"`
+
+      // Value of the variable.
+      // Note: the value will be validated against the schema
+      // of the corresponding VariableDefinitionClass from the ClusterClass.
+      Value apiextensionsv1.JSON `json:"value"`
+    }
+    ```
+
+#### Builtin variables
+
+It’s also possible to use so-called builtin variables in addition to user-defined variables. The following builtin variables are available:
+- `cluster.{name,namespace,labels,annotations}`
+- `cluster.topology.{version,class}`
+- `cluster.topology.controlPlane.{labels,annotations,replicas}`
+    - **Note**: these variables are only available when patching control plane or control plane machine templates.
+- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,class,name}`
+    - **Note**: these variables are only available when patching MachineDeployment templates and contain the values 
+      of the current `MachineDeploymentTopology`.
+
+Builtin variables are available under the `builtin.` prefix. Some examples:
+- Usage of `cluster.name` via variable:
+   ```yaml
+   jsonPatches:
+   - op: replace
+     path: “/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/controllerManager/extraArgs/cluster-name”
+     valueFrom:
+       variable: builtin.cluster.name
+   ```
+- Usage of `cluster.name` via template:
+   ```yaml
+   jsonPatches:
+   - op: replace
+     path: “/spec/template/spec/kubeadmConfigSpec/files”
+     valueFrom:
+       template: |
+         - contentFrom:
+             secret
+               key: control-plane-azure.json
+               name: {{ .builtin.cluster.name }}-control-plane-azure-json
+   ```
 
 #### Validations
+
 ##### ClusterClass
 - For object creation:
   - (defaulting) if namespace field is empty for a reference, default it to `metadata.Namespace`
   - all the reference must be in the same namespace of `metadata.Namespace`
   - `spec.workers.machineDeployments[i].class` field must be unique within a ClusterClass.
+  - `VariableDefinitionClass`:
+    - names must be unique, not empty and not equal to `builtin`
+    - schemas must be valid
```

Comment:
Valid OpenAPI v3 schemas?

Reply:
It's not the same as a OpenAPI v3 schema (not even the one in CRDs matches the spec entirely). 
I would prefer keeping this a bit vague in the proposal as the exact validation is very close to what is done for CRDs and rather complex.

Reply:
Sure - understood. I was just hoping there was a simple answer to this complicated question 😄

Reply:
Unfortunately, there will probably only be one rather long answer in the PR when we implement it :)